### ISSUE TITLE
`mem::size_of_val` returns wrong value for DSTs with certain field layouts

### ISSUE THREAD

**Issue by apasel422** [rust-lang/rust#35815]
On stable, beta, and nightly, the follow code prints `16` followed by `24`:

```rust
struct Foo<T: ?Sized> {
    a: i64,
    b: bool,
    c: T,
}

fn main() {
    let a = Box::new(Foo {a: 1, b: false, c: 1u32});
    println!("{}", std::mem::size_of_val(&*a));

    let a: Box<Foo<Send>> = a;
    println!("{}", std::mem::size_of_val(&*a));
}
```

CC #27023

**Comment by eddyb** [comment]
This is probably for drop flags, in which case it should get fixed by #35764

**Comment by apasel422** [comment]
Reduced:

```rust
struct Foo<T: ?Sized> {
    a: i64,
    b: bool,
    c: T,
}

fn main() {
    let a: &Foo<i32> = &Foo { a: 1, b: false, c: 2i32 };
    println!("{}", std::mem::size_of_val(a));
    
    let a: &Foo<Send> = a;
    println!("{}", std::mem::size_of_val(a));
}
```

**Comment by durka** [comment]
@eddyb where would there be a drop flag here?

**Comment by eddyb** [comment]
@durka My point isn't that there is, but that the code assumes all trait objects have drop flags. I could be wrong though, it might be a bug in the alignment logic.

**Comment by eddyb** [comment]
Not fixed by #35764, will have to investigate further.

**Comment by eddyb** [comment]
Uhhhh, the computation starts with:
```rust
DEBUG:rustc_trans::glue: DST Foo<std::marker::Send> sizing_type: { i64, i8 }
DEBUG:rustc_trans::glue: DST Foo<std::marker::Send> statically sized prefix size: 16 align: 8
```
Which is completely wrong, because you want no trailing padding, i.e. size would be `9`.
